web_monitoring/db.py

```diff
@@ -1,47 +1,608 @@
+    -------
+    capture_time_query : None or string
+        If None, do not query ``capture_time``.
+    """
+    if start_date is None and end_date is None:
+        return None
+    if start_date is not None:
+        start_str = start_date.isoformat() + 'Z'
+    else:
+        start_str = ''
+    if end_date is not None:
+        end_str = end_date.isoformat() + 'Z'
+    else:
+        end_str = ''
+    return f'{start_str}..{end_str}'
+
+
+def parse_timestamp(capture_time):
+    return datetime.strptime(capture_time, '%Y-%m-%dT%H:%M:%S.%fZ')
+
+
+class MissingCredentials(RuntimeError):
+    ...
+
+
+def ensure_credentials(f):
+    """Check that the module settings dict is populated."""
+    @functools.wraps(f)
+    def inner(*args, **kwargs):
+        if None in settings.values():
+            raise MissingCredentials("""
+Before using the function, database credentials must be provided to the module.
+You can do this in one of two ways:
+
+1. Update the dictionary `web_monitoring.db.settings`.
+2. Set the following environment variables and then reload the
+   `web_monitoring.db` module.
+
+   WEB_MONITORING_DB_URL
+   WEB_MONITORING_DB_EMAIL
+   WEB_MONITORING_DB_PASSWORD""")
+        return f(*args, **kwargs)
+    return inner
+
+
+### PAGES ###
+
+
+@ensure_credentials
+def list_pages(chunk=None, chunk_size=None,
+               site=None, agency=None, url=None, title=None,
+               include_versions=None, source_type=None, version_hash=None,
+               start_date=None, end_date=None):
+    """
+    List all Pages, optionally filtered by search criteria.
+
+    Parameters
+    ----------
+    chunk : integer, optional
+        pagination parameter
+    chunk_size : integer, optional
+        number of items per chunk
+    site : string, optional
+    agency : string, optional
+    url : string, optional
+    title : string, optional
+    include_versions : boolean, optional
+    source_type : string, optional
+        Only relevant if ``include_versions`` is True
+    version_hash : string, optional
+        Only relevant if ``include_versions`` is True
+    start_date : datetime, optional
+        Only relevant if ``include_versions`` is True
+    end_date : datetime, optional
+        Only relevant if ``include_versions`` is True
+
+    Returns
+    -------
+    response : dict
+    """
+    params = {'chunk': chunk,
+              'chunk_size': chunk_size,
+              'site': site,
+              'agency': agency,
+              'url': url,
+              'title': title,
+              'include_versions': include_versions}
+    if include_versions is not None:
+        params.update({'source_type': source_type,
+                       'version_hash': version_hash,
+                       'capture_time': time_range_string(start_date,
+                                                          end_date)})
+    url = f'{api_url()}/pages'
+    res = requests.get(url, auth=auth(), params=params)
+    res.raise_for_status()
```

Comment:
Which `errors` property? I don't see it in the [API docs](http://docs.python-requests.org/en/master/api/#requests.Response) or in vars:

```python
In [11]: res = requests.get('https://api.monitoring.envirodatagov.org/api/v0/error')

In [12]: vars(res)
Out[12]: 
{'_content': b'<!DOCTYPE html>\n<html>\n<head>\n  <title>The page you were looking for doesn\'t exist (404)</title>\n  <meta name="viewport" content="width=device-width,initial-scale=1">\n  <style>\n  body {\n    background-color: #EFEFEF;\n    color: #2E2F30;\n    text-align: center;\n    font-family: arial, sans-serif;\n    margin: 0;\n  }\n\n  div.dialog {\n    width: 95%;\n    max-width: 33em;\n    margin: 4em auto 0;\n  }\n\n  div.dialog > div {\n    border: 1px solid #CCC;\n    border-right-color: #999;\n    border-left-color: #999;\n    border-bottom-color: #BBB;\n    border-top: #B00100 solid 4px;\n    border-top-left-radius: 9px;\n    border-top-right-radius: 9px;\n    background-color: white;\n    padding: 7px 12% 0;\n    box-shadow: 0 3px 8px rgba(50, 50, 50, 0.17);\n  }\n\n  h1 {\n    font-size: 100%;\n    color: #730E15;\n    line-height: 1.5em;\n  }\n\n  div.dialog > p {\n    margin: 0 0 1em;\n    padding: 1em;\n    background-color: #F7F7F7;\n    border: 1px solid #CCC;\n    border-right-color: #999;\n    border-left-color: #999;\n    border-bottom-color: #999;\n    border-bottom-left-radius: 4px;\n    border-bottom-right-radius: 4px;\n    border-top-color: #DADADA;\n    color: #666;\n    box-shadow: 0 3px 8px rgba(50, 50, 50, 0.17);\n  }\n  </style>\n</head>\n\n<body>\n  <!-- This file lives in public/404.html -->\n  <div class="dialog">\n    <div>\n      <h1>The page you were looking for doesn\'t exist.</h1>\n      <p>You may have mistyped the address or the page may have moved.</p>\n    </div>\n    <p>If you are the application owner check the logs for more information.</p>\n  </div>\n</body>\n</html>\n',
 '_content_consumed': True,
 '_next': None,
 'connection': <requests.adapters.HTTPAdapter at 0x10eb6ebe0>,
 'cookies': <RequestsCookieJar[]>,
 'elapsed': datetime.timedelta(0, 0, 659360),
 'encoding': 'UTF-8',
 'headers': {'Server': 'Cowboy', 'Date': 'Tue, 10 Oct 2017 14:33:12 GMT', 'Connection': 'keep-alive', 'Content-Type': 'text/html; charset=UTF-8', 'X-Request-Id': 'fca7380c-20fb-4835-ad4e-ebd6642e56ce', 'X-Runtime': '0.002000', 'Strict-Transport-Security': 'max-age=15552000; includeSubDomains', 'Vary': 'Origin', 'Content-Length': '1564', 'Via': '1.1 vegur'},
 'history': [],
 'raw': <urllib3.response.HTTPResponse at 0x10eb633c8>,
 'reason': 'Not Found',
 'request': <PreparedRequest [GET]>,
 'status_code': 404,
 'url': 'https://api.monitoring.envirodatagov.org/api/v0/error'}

In [13]: res.raise_for_status()
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-13-6ec835a6a98f> in <module>()
----> 1 res.raise_for_status()

~/miniconda3/envs/dd/lib/python3.6/site-packages/requests/models.py in raise_for_status(self)
    933 
    934         if http_error_msg:
--> 935             raise HTTPError(http_error_msg, response=self)
    936 
    937     def close(self):

HTTPError: 404 Client Error: Not Found for url: https://api.monitoring.envirodatagov.org/api/v0/error
```

Reply:
Sorry, I mean the API returns meaningful data on errors. e.g. this request: `https://api.monitoring.envirodatagov.org/api/v0/pages?capture_time=2017-07-30T20:00:00.000Z..2017-08-0`

Returns a 400 status code, but it *also* returns JSON data that is more specific and helpful:

```json
{
  "errors": [
    {
      "status": 400,
      "title": "Invalid date: '2017-08-0'"
    }
  ]
}
```

All routes in the API do this, though in a few rare cases (mostly auth, where it was harder to change the underlying system), it’ll be in the format `{"error": {"message": "string"}}`

Reply:
(Note if you are getting HTML in the response, that is because you are in development mode, where the HTML is pretty useful because it includes an interactive interpreter and formatted stack traces. Open to changing that if you don’t think it’s worth keeping.)